File: drug_server.py
#new mcp server for drug features
import os
from dotenv import load_dotenv
import sys
from typing import Dict, Any, List
import asyncio

# Add current directory to Python path to find our client module
sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))

from mcp.server.fastmcp import FastMCP
import drug_features

load_dotenv()

mcp_app = FastMCP(
    name="DrugFeaturesService",
    version="1.0.0",
    description="MCP server for drug interactions, name conversion, and adverse events"
)

@mcp_app.tool()
async def check_drug_interactions(
    drug1: str,
    drug2: str,
    additional_drugs: List[str] = []
) -> Dict[str, Any]:
    """
    Check for potential drug interactions between medications using RxNav API
    
    Args:
        drug1: First medication name
        drug2: Second medication name  
        additional_drugs: Optional list of additional medications to check
    
    Returns:
        Dictionary containing interaction analysis results
    """
    loop = asyncio.get_event_loop()
    interaction_results = await loop.run_in_executor(
        None, 
        drug_features.check_drug_interactions, 
        drug1, 
        drug2, 
        additional_drugs
    )
    
    return {
        "interaction_analysis": interaction_results,
        "data_source": "RxNorm API (ingredient analysis)",
        "analysis_type": "Basic Drug Safety Check",
        "note": "Limited to ingredient comparison - consult pharmacist for comprehensive interaction checking"
    }

@mcp_app.tool()
async def convert_drug_names(
    drug_name: str,
    conversion_type: str = "both"  # "generic", "brand", "both"
) -> Dict[str, Any]:
    """
    Convert between generic and brand names using OpenFDA label data
    
    Args:
        drug_name: Name of the drug to convert
        conversion_type: Type of conversion - "generic", "brand", or "both"
    
    Returns:
        Dictionary containing name conversion results
    """
    loop = asyncio.get_event_loop()
    conversion_results = await loop.run_in_executor(
        None, 
        drug_features.convert_drug_names, 
        drug_name, 
        conversion_type
    )
    
    return {
        "name_conversion": conversion_results,
        "data_source": "openFDA Drug Label API",
        "analysis_type": "Drug Name Conversion",
        "note": "Uses existing FDA labeling data for name mapping"
    }

@mcp_app.tool()
async def get_adverse_events(
    drug_name: str,
    time_period: str = "1year",
    severity_filter: str = "all"  # "all", "serious"
) -> Dict[str, Any]:
    """
    Get FDA adverse event reports for a medication from FAERS database
    
    Args:
        drug_name: Name of the medication
        time_period: Time period for analysis (currently not implemented in API)
        severity_filter: Filter by severity - "all" or "serious" only
    
    Returns:
        Dictionary containing adverse event analysis results
    """
    loop = asyncio.get_event_loop()
    adverse_event_results = await loop.run_in_executor(
        None, 
        drug_features.get_adverse_events, 
        drug_name, 
        time_period, 
        severity_filter
    )
    
    return {
        "adverse_event_analysis": adverse_event_results,
        "data_source": "FDA FAERS (Adverse Event Reporting System)",
        "analysis_type": "Post-Market Safety Surveillance",
        "note": "Real-world adverse event data from healthcare providers and patients"
    }

if __name__ == "__main__":
    print("drug features server starting", file=sys.stderr)
    mcp_app.run(transport='stdio')

File: Local_Version_of_server/JS_version/drug-server.js
#!/usr/bin/env node

// drug-server.js - MCP server for drug features
import dotenv from 'dotenv';
import { Server } from '@modelcontextprotocol/sdk/server/index.js';
import { StdioServerTransport } from '@modelcontextprotocol/sdk/server/stdio.js';
import { 
    CallToolRequestSchema,
    ListToolsRequestSchema,
} from '@modelcontextprotocol/sdk/types.js';
import {
    checkDrugInteractions,
    convertDrugNames,
    getAdverseEvents
} from './drug-features.js';

// Load environment variables from .env file
dotenv.config();

// Create the MCP server
const server = new Server(
    {
        name: "DrugFeaturesService",
        version: "1.0.0",
    },
    {
        capabilities: {
            tools: {},
        },
    }
);

// List available tools
server.setRequestHandler(ListToolsRequestSchema, async () => {
    return {
        tools: [
            {
                name: "check_drug_interactions",
                description: "Check for potential drug interactions between medications using RxNav API",
                inputSchema: {
                    type: "object",
                    properties: {
                        drug1: {
                            type: "string",
                            description: "First medication name"
                        },
                        drug2: {
                            type: "string", 
                            description: "Second medication name"
                        },
                        additional_drugs: {
                            type: "array",
                            items: { type: "string" },
                            description: "Optional list of additional medications to check",
                            default: []
                        }
                    },
                    required: ["drug1", "drug2"]
                }
            },
            {
                name: "convert_drug_names",
                description: "Convert between generic and brand names using OpenFDA label data",
                inputSchema: {
                    type: "object",
                    properties: {
                        drug_name: {
                            type: "string",
                            description: "Name of the drug to convert"
                        },
                        conversion_type: {
                            type: "string",
                            description: "Type of conversion - 'generic', 'brand', or 'both'",
                            enum: ["generic", "brand", "both"],
                            default: "both"
                        }
                    },
                    required: ["drug_name"]
                }
            },
            {
                name: "get_adverse_events",
                description: "Get FDA adverse event reports for a medication from FAERS database",
                inputSchema: {
                    type: "object",
                    properties: {
                        drug_name: {
                            type: "string",
                            description: "Name of the medication"
                        },
                        time_period: {
                            type: "string",
                            description: "Time period for analysis (currently not implemented in API)",
                            default: "1year"
                        },
                        severity_filter: {
                            type: "string",
                            description: "Filter by severity - 'all' or 'serious' only",
                            enum: ["all", "serious"],
                            default: "all"
                        }
                    },
                    required: ["drug_name"]
                }
            }
        ]
    };
});

// Handle tool calls
server.setRequestHandler(CallToolRequestSchema, async (request) => {
    const { name, arguments: args } = request.params;

    try {
        switch (name) {
            case "check_drug_interactions": {
                const { drug1, drug2, additional_drugs = [] } = args;
                
                const interactionResults = await checkDrugInteractions(drug1, drug2, additional_drugs);
                
                const result = {
                    interaction_analysis: interactionResults,
                    data_source: "RxNorm API (ingredient analysis)",
                    analysis_type: "Basic Drug Safety Check",
                    note: "Limited to ingredient comparison - consult pharmacist for comprehensive interaction checking"
                };
                
                return {
                    content: [
                        {
                            type: "text",
                            text: JSON.stringify(result, null, 2)
                        }
                    ]
                };
            }

            case "convert_drug_names": {
                const { drug_name, conversion_type = "both" } = args;
                
                const conversionResults = await convertDrugNames(drug_name, conversion_type);
                
                const result = {
                    name_conversion: conversionResults,
                    data_source: "openFDA Drug Label API",
                    analysis_type: "Drug Name Conversion",
                    note: "Uses existing FDA labeling data for name mapping"
                };
                
                return {
                    content: [
                        {
                            type: "text",
                            text: JSON.stringify(result, null, 2)
                        }
                    ]
                };
            }

            case "get_adverse_events": {
                const { drug_name, time_period = "1year", severity_filter = "all" } = args;
                
                const adverseEventResults = await getAdverseEvents(drug_name, time_period, severity_filter);
                
                const result = {
                    adverse_event_analysis: adverseEventResults,
                    data_source: "FDA FAERS (Adverse Event Reporting System)",
                    analysis_type: "Post-Market Safety Surveillance",
                    note: "Real-world adverse event data from healthcare providers and patients"
                };
                
                return {
                    content: [
                        {
                            type: "text",
                            text: JSON.stringify(result, null, 2)
                        }
                    ]
                };
            }

            default:
                throw new Error(`Unknown tool: ${name}`);
        }
    } catch (error) {
        return {
            content: [
                {
                    type: "text",
                    text: `Error: ${error.message}`
                }
            ],
            isError: true
        };
    }
});

// Start the server
async function main() {
    const transport = new StdioServerTransport();
    await server.connect(transport);
    console.error("Drug Features Service MCP server running on stdio");
}

main().catch((error) => {
    console.error("Fatal error:", error);
    process.exit(1);
});